### PR TITLE
Added models, migrations, endpoints for status, factory type and factory

### PR DESCRIPTION
**Pull request by 1mac11**

### CODE REVIEW

File: core/models/factory.py [hunk 1]
```diff
@@ -0,0 +1,10 @@
+from django.db import models
+
+from core.models import BaseModel
+
+
+class Factory(BaseModel):
+    name = models.CharField(max_length=255)
```
Comment (ahodkov) [inline core/models/factory.py]: unique

File: core/models/factory_type.py [hunk 2]
```diff
@@ -0,0 +1,6 @@
+from django.db import models
+from .base import BaseModel
+
+
+class FactoryType(BaseModel):
+    name = models.CharField(max_length=255)
```
Comment (ahodkov) [inline core/models/factory_type.py]: unique

File: core/models/status.py [hunk 3]
```diff
@@ -0,0 +1,6 @@
+from django.db import models
+from .base import BaseModel
+
+
+class Status(BaseModel):
+    name = models.CharField(max_length=255)
```
Comment (ahodkov) [inline core/models/status.py]: unique

File: core/models/territory.py [hunk 4]
```diff
@@ -9,3 +9,4 @@ class Territory(BaseModel):
     address = models.CharField(max_length=255, blank=True, null=True)
```
Comment (ahodkov) [inline core/models/territory.py]: name should be not null

File: api/views/status.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+from rest_framework.viewsets import ModelViewSet
+
+from api.serializers import StatusSerializer
+from core.models import Status
+
+
+class StatusViewSet(ModelViewSet):
+    queryset = Status.objects.all()
+    serializer_class = StatusSerializer
+    permission_classes = (permissions.AllowAny,)
```
Comment (ahodkov) [inline api/views/status.py]: isAuthenticatedOrReadOnly

File: api/views/factory_type.py [hunk 6]
```diff
@@ -0,0 +1,12 @@
...
+from rest_framework.viewsets import ModelViewSet
+
+from api.serializers import FactoryTypeSerializer
+from core.models import FactoryType
+
+
+class FactoryTypeViewSet(ModelViewSet):
+    queryset = FactoryType.objects.all()
+    serializer_class = FactoryTypeSerializer
+    permission_classes = (permissions.AllowAny,)
```
Comment (ahodkov) [inline api/views/factory_type.py]: isAuthenticatedOrReadOnly

**Review by ahodkov — CHANGES_REQUESTED** [review]
Also, for all serializers you need write custom field like this:
```
fields = ['status']
status = serializers.SerilizersMethodField()
def get_status(self, territory):
     return territory.status.name 

```

File: api/views/factory.py [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+from rest_framework.viewsets import ModelViewSet
+
+from api.serializers import FactorySerializer
+from core.models import Factory
+
+
+class FactoryViewSet(ModelViewSet):
+    queryset = Factory.objects.all()
+    serializer_class = FactorySerializer
+    permission_classes = (permissions.AllowAny,)
```
Comment (ahodkov) [inline api/views/factory.py]: isAuthenticatedOrReadOnly to prevent deleting/updating/creating

**Review by ahodkov — CHANGES_REQUESTED** [review]
Also, I thought about status field. Maybe, good point to use `models.SET_NULL`, because status is additional information and will be bad if we delete PENDING status and delete all building(for example).

File: core/models/machine_tool.py [hunk 8]
```diff
@@ -5,6 +5,7 @@
 
 class MachineTool(BaseModel):
     name = models.CharField(max_length=255, unique=True)
-    type = models.ForeignKey('core.MachineToolType', on_delete=models.CASCADE, related_name='machine_tools')
+    type = models.ForeignKey('core.MachineToolType', on_delete=models.SET_NULL, related_name='machine_tools',
+                             blank=True, null=True)
     status = models.ForeignKey('core.Status', on_delete=models.CASCADE, related_name='machine_tools')
```
Comment (ahodkov) [inline core/models/machine_tool.py]: Change on_delete in status instead of type

**Review by ahodkov — CHANGES_REQUESTED** [review]

File: core/models/user_role.py [hunk 9]
```diff
@@ -0,0 +1,6 @@
+from django.db import models
+from core.models import BaseModel
+
+
+class UserRole(BaseModel):
+    name = models.CharField(max_length=255, unique=True)
```
Comment (ahodkov) [inline core/models/user_role.py]: choices may be better in name

File: core/models/user.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+    use_in_migrations = True
+
+    def _create_user(self, email, password, **extra_fields):
+        """Create and save a User with the given email and password."""
+        if not email:
+            raise ValueError('The given email must be set')
+        email = self.normalize_email(email)
+        user = self.model(email=email, **extra_fields)
+        user.set_password(password)
```
Comment (ahodkov) [inline core/models/user.py]: Need to check: is it hashed in database after saving?

File: core/models/user.py [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+            raise ValueError('Superuser must have is_staff=True.')
+        if extra_fields.get('is_superuser') is not True:
+            raise ValueError('Superuser must have is_superuser=True.')
+
+        return self._create_user(email, password, **extra_fields)
+
+
+class User(AbstractUser):
+    """User model."""
+    username = None
```
Comment (ahodkov) [inline core/models/user.py]: if it's none - it doesn't need

**Review by ahodkov — CHANGES_REQUESTED** [review]

File: core/models/user.py (hunk 10 above)
Comment (1mac11) [reply core/models/user.py]: I think it is hashed. https://docs.djangoproject.com/en/4.0/ref/contrib/auth/#django.contrib.auth.models.User.check_password

File: core/models/user.py (hunk 11 above)
Comment (1mac11) [reply core/models/user.py]: hmm, here I am using django's AbstractUser and username is required so without username = None we can't create user

Comment (ahodkov) [reply core/models/user.py]: So strange, try to use AbstractBaseUser instead

File: core/models/machine_tool.py [hunk 12]
```diff
@@ -0,0 +1,12 @@
+from django.db import models
+
+from core.models import BaseModel
+
+
+class MachineTool(BaseModel):
+    name = models.CharField(max_length=255, unique=True)
+    type = models.ForeignKey('core.MachineToolType', on_delete=models.SET_NULL, related_name='machine_tools',
```
Comment (ahodkov) [inline core/models/machine_tool.py]: `models.CASCADE`

**Review by ahodkov — CHANGES_REQUESTED** [review]
Fix and I will approve

**Review by ahodkov — APPROVED** [review]

**Comment by 1mac11** [comment]
Added models, migrations and endpoints

**Pull request merged**